arch/x86/core/userspace.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys/speculation.h>
+#include <syscall_handler.h>
+#include <kernel_arch_func.h>
+#include <ksched.h>
+
+#ifndef CONFIG_X86_KPTI
+/* Change to new set of page tables. ONLY intended for use from
+ * z_x88_swap_update_page_tables(). This changes CR3, no memory access
+ * afterwards is legal unless it is known for sure that the relevant
+ * mappings are identical wrt supervisor mode until we iret out.
+ */
+static inline void page_tables_set(struct x86_page_tables *ptables)
+{
+#ifdef CONFIG_X86_64
+	__asm__ volatile("movq %0, %%cr3\n\t" : : "r" (ptables) : "memory");
+#else
+	__asm__ volatile("movl %0, %%cr3\n\t" : : "r" (ptables) : "memory");
+#endif
+}
+
+/* Set initial stack pointer for privilege mode elevations */
+static inline void set_initial_psp(char *psp)
+{
+#ifdef CONFIG_X86_64
+	__asm__ volatile("movq %0, %%gs:__x86_tss64_t_psp_OFFSET\n\t"
+			 : : "r" (psp));
+#else
+	_main_tss.esp0 = (uintptr_t)psp;
+#endif
+}
+
+/* Update the to the incoming thread's page table, and update the location of
+ * the privilege elevation stack.
+ *
+ * May be called ONLY during context switch and when supervisor threads drop
+ * synchronously to user mode. Hot code path!
+ *
+ * Nothing to do here if KPTI is enabled. We are in supervisor mode, so the
+ * active page tables are the kernel's page tables. If the incoming thread is
+ * in user mode we are going to switch CR3 to the thread-specific tables when
+ * we go through z_x86_trampoline_to_user.
+ *
+ * We don't need to update the privilege mode initial stack pointer either,
+ * privilege elevation always lands on the trampoline stack and the irq/sycall
+ * code has to manually transition off of it to the thread's kernel stack after
+ * switching page tables.
+ */
+void z_x86_swap_update_page_tables(struct k_thread *incoming)
+{
+	struct x86_page_tables *ptables;
+
+	if ((incoming->base.user_options & K_USER) != 0) {
+		set_initial_psp(incoming->arch.psp);
+	}
+
+	/* Check first that we actually need to do this, since setting
+	 * CR3 involves an expensive full TLB flush.
+	 */
+	ptables = z_x86_thread_page_tables_get(incoming);
+
+	if (ptables != z_x86_page_tables_get()) {
+		page_tables_set(ptables);
+	}
+}
+#endif /* CONFIG_X86_KPTI */
+
+FUNC_NORETURN static void drop_to_user(k_thread_entry_t user_entry,
+				       void *p1, void *p2, void *p3)
+{
+	u32_t stack_end;
+
+	/* Transition will reset stack pointer to initial, discarding
+	 * any old context since this is a one-way operation
+	 */
+	stack_end = STACK_ROUND_DOWN(_current->stack_info.start +
+				     _current->stack_info.size);
+
+	z_x86_userspace_enter(user_entry, p1, p2, p3, stack_end,
+			      _current->stack_info.start);
+	CODE_UNREACHABLE;
+}
+
+/* Does the following:
+ *
+ * - Allows the kernel to write to the privilege elevation stack area.
```

Comment:
this needs to be removed, we don't configure the privilege page as a guard any more. the rest of the comments can be moved inline